### ISSUE TITLE
Image build differences, maybe platform based

### ISSUE THREAD

**Issue by mattyrob** [ClassicPress/ClassicPress#865]
The `grunt precommit:image` command invokes image minification on core images in the wp-admin and wp-includes folders. The process should be platform agnostic and produce the same results.

## Expected behavior
There should be no changes recorded by `git` running this command on the `develop` branch on any platform

## Current behavior
When I run the above command on MacOS (12.0.1) I get changes captures in `git` for 39 images (diff below)
```
diff --git a/src/wp-admin/images/bubble_bg-2x.gif b/src/wp-admin/images/bubble_bg-2x.gif
index 8e34e01dcd..21302a34dc 100644
Binary files a/src/wp-admin/images/bubble_bg-2x.gif and b/src/wp-admin/images/bubble_bg-2x.gif differ
diff --git a/src/wp-admin/images/bubble_bg.gif b/src/wp-admin/images/bubble_bg.gif
index 2888d9a6bc..bcf74c4854 100644
Binary files a/src/wp-admin/images/bubble_bg.gif and b/src/wp-admin/images/bubble_bg.gif differ
diff --git a/src/wp-admin/images/date-button-2x.gif b/src/wp-admin/images/date-button-2x.gif
index 49ff233113..281a6654fa 100644
Binary files a/src/wp-admin/images/date-button-2x.gif and b/src/wp-admin/images/date-button-2x.gif differ
diff --git a/src/wp-admin/images/date-button.gif b/src/wp-admin/images/date-button.gif
index 66ba110028..326a7b71ca 100644
Binary files a/src/wp-admin/images/date-button.gif and b/src/wp-admin/images/date-button.gif differ
diff --git a/src/wp-admin/images/loading.gif b/src/wp-admin/images/loading.gif
index 12d31e3b49..3778ebceb6 100644
Binary files a/src/wp-admin/images/loading.gif and b/src/wp-admin/images/loading.gif differ
diff --git a/src/wp-admin/images/media-button-image.gif b/src/wp-admin/images/media-button-image.gif
index 56a974731d..69efa7b484 100644
Binary files a/src/wp-admin/images/media-button-image.gif and b/src/wp-admin/images/media-button-image.gif differ
diff --git a/src/wp-admin/images/media-button-music.gif b/src/wp-admin/images/media-button-music.gif
index 8fcd3530eb..8e9a30aab2 100644
Binary files a/src/wp-admin/images/media-button-music.gif and b/src/wp-admin/images/media-button-music.gif differ
diff --git a/src/wp-admin/images/media-button-other.gif b/src/wp-admin/images/media-button-other.gif
index 4a19efd3f7..2a7c35dc10 100644
Binary files a/src/wp-admin/images/media-button-other.gif and b/src/wp-admin/images/media-button-other.gif differ
diff --git a/src/wp-admin/images/media-button-video.gif b/src/wp-admin/images/media-button-video.gif
index 405083b0b3..9b4a7963ea 100644
Binary files a/src/wp-admin/images/media-button-video.gif and b/src/wp-admin/images/media-button-video.gif differ
diff --git a/src/wp-admin/images/wpspin_light-2x.gif b/src/wp-admin/images/wpspin_light-2x.gif
index 25a6254bf9..fe5dea6290 100644
Binary files a/src/wp-admin/images/wpspin_light-2x.gif and b/src/wp-admin/images/wpspin_light-2x.gif differ
diff --git a/src/wp-admin/images/wpspin_light.gif b/src/wp-admin/images/wpspin_light.gif
index 7c0c698a23..a080e51f49 100644
Binary files a/src/wp-admin/images/wpspin_light.gif and b/src/wp-admin/images/wpspin_light.gif differ
diff --git a/src/wp-admin/images/xit-2x.gif b/src/wp-admin/images/xit-2x.gif
index 2415429207..040107a90f 100644
Binary files a/src/wp-admin/images/xit-2x.gif and b/src/wp-admin/images/xit-2x.gif differ
diff --git a/src/wp-admin/images/xit.gif b/src/wp-admin/images/xit.gif
index bf9452e5f9..2790d2113f 100644
Binary files a/src/wp-admin/images/xit.gif and b/src/wp-admin/images/xit.gif differ
diff --git a/src/wp-includes/images/smilies/icon_arrow.gif b/src/wp-includes/images/smilies/icon_arrow.gif
index ce6f1f6216..9f7976c615 100644
Binary files a/src/wp-includes/images/smilies/icon_arrow.gif and b/src/wp-includes/images/smilies/icon_arrow.gif differ
diff --git a/src/wp-includes/images/smilies/icon_biggrin.gif b/src/wp-includes/images/smilies/icon_biggrin.gif
index 2752870a20..c994014ad9 100644
Binary files a/src/wp-includes/images/smilies/icon_biggrin.gif and b/src/wp-includes/images/smilies/icon_biggrin.gif differ
diff --git a/src/wp-includes/images/smilies/icon_confused.gif b/src/wp-includes/images/smilies/icon_confused.gif
index 017ff881d1..c8b5f6a75d 100644
Binary files a/src/wp-includes/images/smilies/icon_confused.gif and b/src/wp-includes/images/smilies/icon_confused.gif differ
diff --git a/src/wp-includes/images/smilies/icon_cool.gif b/src/wp-includes/images/smilies/icon_cool.gif
index 16c568f7ea..e04af62a92 100644
Binary files a/src/wp-includes/images/smilies/icon_cool.gif and b/src/wp-includes/images/smilies/icon_cool.gif differ
diff --git a/src/wp-includes/images/smilies/icon_cry.gif b/src/wp-includes/images/smilies/icon_cry.gif
index 5cdba7f42c..03b2ffdc96 100644
Binary files a/src/wp-includes/images/smilies/icon_cry.gif and b/src/wp-includes/images/smilies/icon_cry.gif differ
diff --git a/src/wp-includes/images/smilies/icon_eek.gif b/src/wp-includes/images/smilies/icon_eek.gif
index ac1865a3bf..e21f3e095a 100644
Binary files a/src/wp-includes/images/smilies/icon_eek.gif and b/src/wp-includes/images/smilies/icon_eek.gif differ
diff --git a/src/wp-includes/images/smilies/icon_evil.gif b/src/wp-includes/images/smilies/icon_evil.gif
index 428e56522d..8998434ba4 100644
Binary files a/src/wp-includes/images/smilies/icon_evil.gif and b/src/wp-includes/images/smilies/icon_evil.gif differ
diff --git a/src/wp-includes/images/smilies/icon_exclaim.gif b/src/wp-includes/images/smilies/icon_exclaim.gif
index f8e34ac66e..1e355d6cdc 100644
Binary files a/src/wp-includes/images/smilies/icon_exclaim.gif and b/src/wp-includes/images/smilies/icon_exclaim.gif differ
diff --git a/src/wp-includes/images/smilies/icon_idea.gif b/src/wp-includes/images/smilies/icon_idea.gif
index 09b2353c28..3ce797646d 100644
Binary files a/src/wp-includes/images/smilies/icon_idea.gif and b/src/wp-includes/images/smilies/icon_idea.gif differ
diff --git a/src/wp-includes/images/smilies/icon_lol.gif b/src/wp-includes/images/smilies/icon_lol.gif
index f9ea0715fc..3612734615 100644
Binary files a/src/wp-includes/images/smilies/icon_lol.gif and b/src/wp-includes/images/smilies/icon_lol.gif differ
diff --git a/src/wp-includes/images/smilies/icon_mad.gif b/src/wp-includes/images/smilies/icon_mad.gif
index f42f613fbb..5f233d7130 100644
Binary files a/src/wp-includes/images/smilies/icon_mad.gif and b/src/wp-includes/images/smilies/icon_mad.gif differ
diff --git a/src/wp-includes/images/smilies/icon_mrgreen.gif b/src/wp-includes/images/smilies/icon_mrgreen.gif
index 92d9c4443b..7bc5c0a42b 100644
Binary files a/src/wp-includes/images/smilies/icon_mrgreen.gif and b/src/wp-includes/images/smilies/icon_mrgreen.gif differ
diff --git a/src/wp-includes/images/smilies/icon_neutral.gif b/src/wp-includes/images/smilies/icon_neutral.gif
index 48ab282dc2..3ab93cd696 100644
Binary files a/src/wp-includes/images/smilies/icon_neutral.gif and b/src/wp-includes/images/smilies/icon_neutral.gif differ
diff --git a/src/wp-includes/images/smilies/icon_question.gif b/src/wp-includes/images/smilies/icon_question.gif
index d23fbc7eb8..5a595ba9c2 100644
Binary files a/src/wp-includes/images/smilies/icon_question.gif and b/src/wp-includes/images/smilies/icon_question.gif differ
diff --git a/src/wp-includes/images/smilies/icon_razz.gif b/src/wp-includes/images/smilies/icon_razz.gif
index 77ad1e2295..5bbfd95e73 100644
Binary files a/src/wp-includes/images/smilies/icon_razz.gif and b/src/wp-includes/images/smilies/icon_razz.gif differ
diff --git a/src/wp-includes/images/smilies/icon_redface.gif b/src/wp-includes/images/smilies/icon_redface.gif
index d41e7f00e0..2adb4f2556 100644
Binary files a/src/wp-includes/images/smilies/icon_redface.gif and b/src/wp-includes/images/smilies/icon_redface.gif differ
diff --git a/src/wp-includes/images/smilies/icon_rolleyes.gif b/src/wp-includes/images/smilies/icon_rolleyes.gif
index b0bc06a505..9f83242cd3 100644
Binary files a/src/wp-includes/images/smilies/icon_rolleyes.gif and b/src/wp-includes/images/smilies/icon_rolleyes.gif differ
diff --git a/src/wp-includes/images/smilies/icon_sad.gif b/src/wp-includes/images/smilies/icon_sad.gif
index 7fe84e9fdb..75f2a41944 100644
Binary files a/src/wp-includes/images/smilies/icon_sad.gif and b/src/wp-includes/images/smilies/icon_sad.gif differ
diff --git a/src/wp-includes/images/smilies/icon_smile.gif b/src/wp-includes/images/smilies/icon_smile.gif
index 148ca0d2c6..3abf3b2213 100644
Binary files a/src/wp-includes/images/smilies/icon_smile.gif and b/src/wp-includes/images/smilies/icon_smile.gif differ
diff --git a/src/wp-includes/images/smilies/icon_surprised.gif b/src/wp-includes/images/smilies/icon_surprised.gif
index 601dfa9694..b84fcb42b0 100644
Binary files a/src/wp-includes/images/smilies/icon_surprised.gif and b/src/wp-includes/images/smilies/icon_surprised.gif differ
diff --git a/src/wp-includes/images/smilies/icon_twisted.gif b/src/wp-includes/images/smilies/icon_twisted.gif
index 535c10f1f4..04457db356 100644
Binary files a/src/wp-includes/images/smilies/icon_twisted.gif and b/src/wp-includes/images/smilies/icon_twisted.gif differ
diff --git a/src/wp-includes/images/smilies/icon_wink.gif b/src/wp-includes/images/smilies/icon_wink.gif
index 8edacae8fb..ab04b93c0e 100644
Binary files a/src/wp-includes/images/smilies/icon_wink.gif and b/src/wp-includes/images/smilies/icon_wink.gif differ
diff --git a/src/wp-includes/images/wpspin-2x.gif b/src/wp-includes/images/wpspin-2x.gif
index 25a6254bf9..fe5dea6290 100644
Binary files a/src/wp-includes/images/wpspin-2x.gif and b/src/wp-includes/images/wpspin-2x.gif differ
diff --git a/src/wp-includes/images/wpspin.gif b/src/wp-includes/images/wpspin.gif
index 7c0c698a23..a080e51f49 100644
Binary files a/src/wp-includes/images/wpspin.gif and b/src/wp-includes/images/wpspin.gif differ
diff --git a/src/wp-includes/images/xit-2x.gif b/src/wp-includes/images/xit-2x.gif
index 2415429207..040107a90f 100644
Binary files a/src/wp-includes/images/xit-2x.gif and b/src/wp-includes/images/xit-2x.gif differ
diff --git a/src/wp-includes/images/xit.gif b/src/wp-includes/images/xit.gif
index bf9452e5f9..2790d2113f 100644
Binary files a/src/wp-includes/images/xit.gif and b/src/wp-includes/images/xit.gif differ
```

## Possible solution
Investigate alternative image minification settings or npm packages that avoid this issue.

## Steps to reproduce (for bugs)
1. You will probably need a MacOS or Windows machine for testing
2. Checkout the code and run `npm install`
3. Run `grunt precommit:image`
4. Run `git status` or `git diff` to see if changes are reported

## Context
When working on ClassicPress and using the `precommit` steps, I have to discard changes to the images before making a commi and push.

**Comment by mattyrob** [comment]
I've been looking into this a little bit more and have identified one possible workaround, but it seems it would require an upstream fix at `grunt-contrib-imagemin`, or for us to fork this done for our own use.

`grunt-contrib-imagemin` depends on `imagemin` and it seems there are reported (and perhaps expected) binary differences in the file output on different platforms:
https://github.com/imagemin/imagemin/issues/250

I have locally tested and all of the images are changes locally with either the same file size afterwards or in some cases the file increases in size:
```
xxd -c 1 file1 | cut -d ' ' -f 2 > file1.hex
xxd -c 1 file2 | cut -d ' ' -f 2 > file2.hex

diff file1.hex file2.hex
```

Running ` grunt precommit:image --verbose` locally reports every image is 'already opitimzed' and yet some files are changed.
In fact the `node_modules/grunt-contrib-imagemin/tasks/imagemin.js` file overwrites all images, even it is is the same size of larger with the following code:

```
if (saved > 0) {
	totalBytes += originalSize;
	totalSavedBytes += saved;
	totalFiles++;
}

grunt.file.write(file.dest, optimizedBuf);
grunt.verbose.writeln(chalk.green('✔ ') + file.src[0] + chalk.gray(` (${msg})`));
```

If we create our own fork with this simple change, this change amends current behaviour to ensure that images of the same size or larger size images are not re-written. That should improve efficiency locally as well as avoiding creation of larger images. This change also stops this being an issue for me locally when running `grunt precommit`.

```
if (saved > 0) {
	totalBytes += originalSize;
	totalSavedBytes += saved;
	totalFiles++;
	grunt.file.write(file.dest, optimizedBuf);
	grunt.verbose.writeln(chalk.green('✔ ') + file.src[0] + chalk.gray(` (${msg})`));
}
```

**Comment by viktorix** [comment]
@mattyrob still an issue with v2?

**Comment by mattyrob** [comment]
Fixed with #1346
